modules/sr/robot3/arduino_devices.py

```diff
@@ -0,0 +1,234 @@
+from __future__ import annotations
+
+import abc
+import enum
+import random
+import logging
+
+from controller import (
+    LED,
+    Robot,
+    TouchSensor,
+    DistanceSensor as WebotsDistanceSensor,
+)
+from sr.robot3.utils import map_to_range, get_robot_device
+from sr.robot3.randomizer import add_jitter
+from sr.robot3.output_frequency_limiter import OutputFrequencyLimiter
+
+LOGGER = logging.getLogger(__name__)
+
+
+class GPIOPinMode(str, enum.Enum):
+    """The possible modes for a GPIO pin."""
+    INPUT = 'INPUT'
+    INPUT_PULLUP = 'INPUT_PULLUP'
+    OUTPUT = 'OUTPUT'
+
+
+DIGITAL_READ_MODES = {GPIOPinMode.INPUT, GPIOPinMode.INPUT_PULLUP, GPIOPinMode.OUTPUT}
+DIGITAL_WRITE_MODES = {GPIOPinMode.OUTPUT}
+ANALOG_READ_MODES = {GPIOPinMode.INPUT}
+
+
+class PinDevice(abc.ABC):
```

Comment:
The repr from the kit seems to be missing.

Reply:
c6e8d59232e73feec807ff1bc4a11e382a96c296

Reply:
This still seems to lack the pin number.

Reply:
3769379